jetty-server/src/main/java/org/eclipse/jetty/server/BlockingContentProducer.java

```diff
@@ -0,0 +1,142 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.concurrent.Semaphore;
+
+/**
+ * Blocking implementation of {@link ContentProducer}. Calling {@link #nextContent()} will block when
+ * there is no available content but will never return null.
+ */
+class BlockingContentProducer implements ContentProducer
+{
+    private final Semaphore _semaphore = new Semaphore(0);
+    private final AsyncContentProducer _asyncContentProducer;
+
+    public BlockingContentProducer(AsyncContentProducer delegate)
+    {
+        _asyncContentProducer = delegate;
+    }
+
+    @Override
+    public void recycle()
+    {
+        _asyncContentProducer.recycle();
+        _semaphore.drainPermits();
+    }
+
+    @Override
+    public int available()
+    {
+        return _asyncContentProducer.available();
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        return _asyncContentProducer.hasContent();
+    }
+
+    @Override
+    public boolean isError()
+    {
+        return _asyncContentProducer.isError();
+    }
+
+    @Override
+    public void checkMinDataRate()
+    {
+        _asyncContentProducer.checkMinDataRate();
+    }
+
+    @Override
+    public long getRawContentArrived()
+    {
+        return _asyncContentProducer.getRawContentArrived();
+    }
+
+    @Override
+    public boolean consumeAll(Throwable x)
+    {
+        return _asyncContentProducer.consumeAll(x);
+    }
+
+    @Override
+    public HttpInput.Content nextContent()
+    {
+        while (true)
+        {
+            HttpInput.Content content = _asyncContentProducer.nextContent();
+            if (content != null)
+                return content;
+
+            if (_asyncContentProducer.isReady())
+                continue;
+
```

Comment:
A comment here to say that IFF isReady() returns false has `needContent` been called, thus we no that eventually a call to `onContentProducible` will come.

Reply:
Done.